README.md
# Statement

***This repository is for personal study and research purposes only.***  
**Please DO NOT USE IT FOR COMMERCIAL PURPOSES.**  
***Violators are responsible for their own.***  
**☣REFUSE to SPREAD☣**
***Thank you for your cooperation.***

## [★](./Dict) Dictionaries

### Learning-Markdown

> [README](https://github.com/Oreomeow/README#readme)  
> [GFM](https://592592.xyz/2020/Github_Flavored_Markdown/#%E7%9B%AE%E5%BD%95)

### Docker

> [Docker Tutorials | Rookie Tutorials](https://www.runoob.com/docker/docker-tutorial.html)  

## [★](./Conf) Configurations  

- [Env](./Conf#readme)
- [config.sample.sh](./Conf/Qinglong/config.sample.sh)

## [★](./Tasks) Tasks

### [<img src="/Icons/qinglong/QL.png" title="QL" width="30" height="30" />][QL] ← More

```sh
ql repo https://github.com/shufflewzc/faker2.git "jd_|jx_|gua_|jddj_|getJDCookie" "activity|backUp|Coupon|update" "^jd[^_]|USER|utils|function|^JS|^TS|^JDJRValidator_Pure|^ZooFaker|^sign|ql"
```

- [extra.sh](./Tasks/qlrepo/extra.sh)

## [★](./INS) Instructions

### JD

- [qinglong](./INS/JD/qinglong#readme)
- [nevinee](./INS/JD/nevinee#readme)  

## ☆ QuantumultX

- **Lazy Configurations** *Created by sngxpro & Maintained by me*

  ``` text
  https://raw.githubusercontent.com/Oreomeow/QuanX/master/sngx2021.conf
  ```

- **Allinone TaskSub-Links** *Created by sngxpro & Maintained by me*

  ```text
  https://ghproxy.com/https://raw.githubusercontent.com/Oreomeow/QuanX/master/task/AllinOne.json
  ```

### Summary

- [BoxJs: Sub-Links](./QX/BoxJs#readme)  

## [★](./Scripts) Scripts

### Repositories [☛](./Scripts#repositories)

- [whyour/qinglong](https://github.com/whyour/qinglong)

> Scheduled task control panel for python and javaScript

[QL]:./Tasks/qlrepo#readme


Log/News.md
# News

## 07.31 16:15

* qinglong docker-compose.yml

```yaml
version: "3"
services:
  qinglong:
    image: whyour/qinglong:latest
    container_name: qinglong
    hostname: qinglong
    restart: unless-stopped
    tty: true
    ports:
      - "5700:5700"
      - "5701:5701"
    volumes:
      - ./config:/ql/config
      - ./db:/ql/db
      - ./log:/ql/log
      - ./repo:/ql/repo
      - ./raw:/ql/raw
      - ./scripts:/ql/scripts
      - ./jbot:/ql/jbot
      - ./ninja:/ql/ninja
    environment: 
      - ENABLE_HANGUP=true
      - ENABLE_TG_BOT=true
      - ENABLE_WEB_PANEL=true
      - TZ=CST-8
```

* 解决青龙拉取`smiek2221/scripts`更新错误问题

```sh
docker exec -it qinglong bash -c 'cd repo/smiek2221_scripts && git fetch --all && git reset --hard origin/master && git pull && ql repo https://github.com/smiek2221/scripts.git "jd_" "USER_AGENTS" "JDJRValidator_Pure|sign_graphics_validate|ZooFaker_Necklace"'
```

* `JDHelloWorld` ts

```sh
cd /ql/scripts/
ql repo https://github.com/JDHelloWorld/jd_scripts.git "
jd_|jx_|getJDCookie" "activity|backUp|Coupon|enen" "^jd[^_]|USER"
cp /ql/repo/JDHelloWorld_jd_scripts/package.json .
npm i
npm i -g ts-node typescript @types/node date-fns axios
tsc JDHelloWorld_jd_scripts_jd_cfd.ts
task JDHelloWorld_jd_scripts_jd_cfd.js now
```

## 07.01 18:40

* 解决青龙拉取`JDHelloWorld/jd_scripts`更新错误问题

> <https://github.com/JDHelloWorld/jd_scripts/issues/57>

```sh
docker exec -it qinglong bash -c 'cd repo/JDHelloWorld_jd_scripts && git fetch --all && git reset --hard origin/main && git pull && ql repo https://github.com/JDHelloWorld/jd_scripts.git "jd_|jx_|getJDCookie" "activity|backUp|jd_delCoupon" "^jd[^_]|USER"'
```

## 06.30 16:20

* V4/🌸 等面板隐患

算是给那些正在使用面板(网页版，例：<http://127.0.0.1:5678> )以及镜像作者关掉面板功能后，自己又想方设法重新打开面板的docker使用者的一个忠告吧

⚠️⚠️服务器IP: 端口/shell 可以直接打开终端。例：<http://127.0.0.1:5678/shell>

图一 网图，不是我写的

![1][1]

图二 有人做的小实验。

![2][2]

图三 网传10万京东cookie

![3][3]

京东cookie泄漏危害：  
轻者：拿cookie助力。  
重者：用你的账户下单，花你的红包，小额支付(300以内，下单后可删除订单和收货地址）。  
再不济拿收货地址和订单记录打包成压缩包一条0.3-3元不等拿去卖也是可以的。  
其他危害：自己想🤔

## 06.28 20:30

* 神秘大佬备份

```sh
docker run -d --name jd_cookie -p 6789:6789 -e QYWX_KEY={QYWX_KEY} -e QYWX_AM={QYWX_AM} -e UPDATE_API={UPDATE_API} tonglin138/jd_cookie sls deploy
```

* 导出的镜像文件

> <https://wwr.lanzoui.com/iBCLpqsbate>

## 06.28 18:05

* JDC fix remote scan 安装方法

终端输入

```sh
yum install wget unzip -y
```

```sh
cd /root 
```

```sh
ls -l 
```

**AMD64 终端输入**

```sh
wget https://github.com/Zy143L/jdc/releases/download/2.0/JDC
```

**arm64 终端输入**

```sh
wget https://github.com/Zy143L/jdc/releases/download/2.0/JDC_arm64
```

终端输入

```sh
chmod 777 JDC 

./JDC 
```

修改 `config.toml` 中的 `path` 项为 `ql`

**AMD64 终端输入**

```sh
nohup ./JDC 1>/dev/null 2>&1 & #AMD64
```

**arm64 终端输入**

```sh
nohup ./JDC_arm64 1>/dev/null 2>&1 & #ARM64
```

打开 `http://ip:5701/info` 看到 “JDC is Already！” 即说明安装成功！

前端安装 终端输入

```sh
cd public 
```

```sh
wget http://nm66.top/dist.zip && unzip dist.zip
```

然后直接访问 IP + 5701 即可看到面板。

如果要重新安装 先终端输入

```sh
netstat -lnp|grep 5701
```

比如输出了 tcp 0 0 0.0.0.0:5701 0.0.0.0:* LISTEN 28937/java

然后执行 kill 关闭该应用程序 数字就是后面的 28937

终端输入

```sh
kill -9 28937
```

JDC 修复编译来自网络资源

宝塔需要在安全组打开 5701 端口

具体需要自测

* 神秘大佬

```sh
docker pull lxk0301/jd_cookie
```

```sh
docker run -d --name jd_cookie -p 6789:6789 -e QYWX_KEY={QYWX_KEY} -e QYWX_AM={QYWX_AM} -e UPDATE_API={UPDATE_API} echowxsy/jd_cookie sls deploy
```

## 06.28 17:25

* 青龙面板小工具 适用于 2.2 面板
  * 修复远程扫码问题 支持 AMD64 和 ARM64 架构 *by Zy143L*  
    <https://github.com/Zy143L/jdc>

  * 配套前端地址  
    <https://github.com/Zy143L/JDC_WEB>

  * 配套青龙 2.2-066 不升级版本  
    <https://hub.docker.com/r/limoe/qinglong>

## 06.28 01:30

* 青龙面板无法打开解决命令

```sh
docker exec -it qinglong nginx -c /etc/nginx/nginx.conf
```

* A1/shuye/V3/V4 京东扫码恢复及网页扫码获取

> <https://shimo.im/docs/RkkWrrQVpxk3TTjR>

## 06.27 19:44

* qinglong 2.8 一键添加内部互助功能  
  * 容器内执行 或 docker 宿主机执行
  * 默认仓库为`JDHelloWorld`，使用其他仓库需自行修改

```sh
bash -c "$(curl -fsSL https://raw.githubusercontent.com/liuqitoday/qinglong-auto-sharecodes/master/one-key.sh)"
```

```sh
docker exec -it qinglong bash -c "$(curl -fsSL https://raw.githubusercontent.com/liuqitoday/qinglong-auto-sharecodes/master/one-key.sh)"
```

## 06.26 21:30

* 口袋书店入口  
京东 APP： `首页` -（中间第二页） `京东图书` -（右上角） `签` -（页面中间）每日参加活动

## 06.26 14:50

* 京东查黑号最新办法  

> <https://vip.m.jd.com/scoreDetail/current>

电脑或者手机浏览器复制打开网址，从里面查找 `creditLevel` ，最高 11 级，低于 8 的抢啥都很难，低于 5 的就不要抢购了。需要登陆过京东账户。

## 06.26 14:00

* `panghu999`宠汪汪报错原因：据说国外 vps 才会报这个错误

```js
❗️
宠汪汪, 错误!
  evalmachine. < anonymous >: 1 <
  html >
  ^

  SyntaxError: Unexpected token '<'
at new Script(vm.js: 101: 7)
at createScript(vm.js: 262: 10)
at Object.runInContext(vm.js: 293: 10)
at IncomingMessage. < anonymous > (/ql/scripts / panghu999_jd_scripts_jd_joy.js: 2387: 28)
at IncomingMessage.emit(events.js: 388: 22)
at endReadableNT(internal / streams / readable.js: 1336: 12)
at processTicksAndRejections(internal / process / task_queues.js: 82: 21)
```

* `JDHelloWorld`拉取仓库命令如下，cron 设为`55 * * * *`（晚于`panghu999`的`50 * * * *`)

```sh
ql repo https://github.com/JDHelloWorld/jd_scripts "jd_|jx_|getJDCookie" "activity|backUp|jd_delCoupon" "^jd[^_]|USER"
```

```sh
55 * * * *
```

* 宠汪汪同样使用`JDHelloWorld`的仓库，禁用其他的只跑`宠汪汪二代目`即可，cron 改为`15 0-23/2 * * *`
* 宠汪汪兑换使用`JDHelloWorld`的仓库，禁用其他的只跑`宠汪汪兑换二代目`即可，cron 改为`0 0-16/8 * * *`或 ? `57-59/1 59 7-23/8 * * *`（自行测试）
* `JDHelloWorld`宠汪汪验证解决命令

```sh
docker exec -it qinglong pnpm i png-js -S
```

* `panghu999`宠汪汪验证解决命令

```sh
docker exec -it qinglong /bin/sh
```

```sh
apk add --no-cache build-base g++ cairo-dev pango-dev giflib-dev
```

```sh
cd scripts
```

```sh
npm install canvas --build-from-source
```

```sh
exit
```

[1]:https://github.com/Oreomeow/VIP/blob/main/Icons/News/1.jpg  
[2]:https://github.com/Oreomeow/VIP/blob/main/Icons/News/2.jpg
[3]:https://github.com/Oreomeow/VIP/blob/main/Icons/News/3.jpg


Scripts/sh/jrd1.sh
#!/usr/bin/env bash
## Date: 2021-04-08
## 没错，如果你打开脚本来看就会发现我就是用Ctrl+C, Ctrl+V 大法从lkx0301 大佬 和 e大佬那里抄过来的，版权归他们
##

IFS='' read -r -d '' lrr <<"EOF"
/*
只能用一天。需每天寻找直播抓包
1 0-23 * * * jrd1.js
*/
const $ = new Env('整点京豆雨');
let allMessage = '';
let bodyList = {
  '20': {
    url: 'https://api.m.jd.com/client.action?functionId=liveActivityV842&uuid=8888888&client=apple&clientVersion=9.4.4&st=1616761511931&sign=1c8a8ab69f27e2bb7b8539dfa5ec3a78&sv=100',
    body: 'body=%7B%22liveId%22:%223759184%22%7D'
  }
}
let ids = {
  '0': 'RRAID',
  '1': 'RRAID',
  '2': 'RRAID',
  '3': 'RRAID',
  '4': 'RRAID',
  '5': 'RRAID',
  '6': 'RRAID',
  '7': 'RRAID',
  '8': 'RRAID',
  '9': 'RRAID',
  '10': 'RRAID',
  '11': 'RRAID',
  '12': 'RRAID',
  '13': 'RRAID',
  '14': 'RRAID',
  '15': 'RRAID',
  '16': 'RRAID',
  '17': 'RRAID',
  '18': 'RRAID',
  '19': 'RRAID',
  '20': 'RRAID',
  '21': 'RRAID',
  '22': 'RRAID',
  '23': 'RRAID'
}
const notify = $.isNode() ? require('./sendNotify') : '';
//Node.js用户请在jdCookie.js处填写京东ck;
const jdCookieNode = $.isNode() ? require('./jdCookie.js') : '';
//IOS等用户直接用NobyDa的jd cookie
let cookiesArr = [], cookie = '', message;
if ($.isNode()) {
  Object.keys(jdCookieNode).forEach((item) => {
    cookiesArr.push(jdCookieNode[item])
  })
  if (process.env.JD_DEBUG && process.env.JD_DEBUG === 'false') console.log = () => {
  };
  if (JSON.stringify(process.env).indexOf('GITHUB') > -1) process.exit(0)
} else {
  cookiesArr = [$.getdata('CookieJD'), $.getdata('CookieJD2'), ...jsonParse($.getdata('CookiesJD') || "[]").map(item => item.cookie)].filter(item => !!item);
}
const JD_API_HOST = 'https://api.m.jd.com/api';
!(async () => {
  if (!cookiesArr[0]) {
    $.msg($.name, '【提示】请先获取京东账号一cookie\n直接使用NobyDa的京东签到获取', 'https://bean.m.jd.com/', {"open-url": "https://bean.m.jd.com/"});
    return;
  }
  await getRedRain();

  let nowTs = new Date().getTime()
  // if (!($.st <= nowTs && nowTs < $.ed)) {
  $.log(`远程红包雨配置获取错误，从本地读取配置`)
  let hour = (new Date().getUTCHours() + 8) % 24
  if (ids[hour]) {
    $.activityId = ids[hour]
    $.log(`本地红包雨配置获取成功`)
  } else {
    $.log(`无法从本地读取配置，请检查运行时间`)
    return
  }
  // } else{
  //   $.log(`远程红包雨配置获取成功`)
  // }
  for (let i = 0; i < cookiesArr.length; i++) {
    if (cookiesArr[i]) {
      cookie = cookiesArr[i];
      $.UserName = decodeURIComponent(cookie.match(/pt_pin=([^; ]+)(?=;?)/) && cookie.match(/pt_pin=([^; ]+)(?=;?)/)[1])
      $.index = i + 1;
      $.isLogin = true;
      $.nickName = '';
      message = '';
      await TotalBean();
      console.log(`\n******开始【京东账号${$.index}】${$.nickName || $.UserName}*********\n`);
      if (!$.isLogin) {
        $.msg($.name, `【提示】cookie已失效`, `京东账号${$.index} ${$.nickName || $.UserName}\n请重新登录获取\nhttps://bean.m.jd.com/`, {"open-url": "https://bean.m.jd.com/"});

        if ($.isNode()) {
          await notify.sendNotify(`${$.name}cookie已失效 - ${$.UserName}`, `京东账号${$.index} ${$.UserName}\n请重新登录获取cookie`);
        }
        continue
      }
      let nowTs = new Date().getTime() + new Date().getTimezoneOffset() * 60 * 1000 + 8 * 60 * 60 * 1000
      console.log(nowTs, $.startTime, $.endTime, cookie)
      await receiveRedRain();
      // await showMsg();
    }
  }
  if (allMessage) {
    if ($.isNode()) await notify.sendNotify(`${$.name}`, `${allMessage}`);
    $.msg($.name, '', allMessage);
  }
})()
    .catch((e) => {
      $.log('', `❌ ${$.name}, 失败! 原因: ${e}!`, '')
    })
    .finally(() => {
      $.done();
    })

function showMsg() {
  return new Promise(resolve => {
    $.msg($.name, '', `【京东账号${$.index}】${$.nickName}\n${message}`);
    resolve()
  })
}

function getRedRain() {
  let body
  if (bodyList.hasOwnProperty(new Date().getDate())) {
    body = bodyList[new Date().getDate()]
  } else {
    return
  }
  return new Promise(resolve => {
    $.post(taskGetUrl(body.url, body.body), (err, resp, data) => {
      try {
        if (err) {
          console.log(`${JSON.stringify(err)}`)
          console.log(`${$.name} API请求失败，请检查网路重试`)
        } else {
          if (safeGet(data)) {
            data = JSON.parse(data);
            if (data.data && data.data.iconArea) {
              console.log(data.data.iconArea.filter(vo => vo['type'] === 'anchor_darw_lottery').length &&
                  data.data.iconArea.filter(vo => vo['type'] === 'anchor_darw_lottery')[0].data.lotteryId)
              let act = data.data.iconArea.filter(vo => vo['type'] === "platform_red_packege_rain")[0]
              if (act) {
                let url = act.data.activityUrl
                $.activityId = url.substr(url.indexOf("id=") + 3)
                $.st = act.startTime
                $.ed = act.endTime
                console.log($.activityId)

                console.log(`下一场红包雨开始时间：${new Date($.st)}`)
                console.log(`下一场红包雨结束时间：${new Date($.ed)}`)
              } else {
                console.log(`暂无红包雨`)
              }
            } else {
              console.log(`暂无红包雨`)
            }
          }
        }
      } catch (e) {
        $.logErr(e, resp)
      } finally {
        resolve();
      }
    })
  })
}

function receiveRedRain() {
  return new Promise(resolve => {
    const body = {"actId": $.activityId};
    $.get(taskUrl('noahRedRainLottery', body), (err, resp, data) => {
      try {
        if (err) {
          console.log(`${JSON.stringify(err)}`)
          console.log(`${$.name} API请求失败，请检查网路重试`)
        } else {
          if (safeGet(data)) {
            data = JSON.parse(data);
            if (data.subCode === '0') {
              console.log(`领取成功，获得${JSON.stringify(data.lotteryResult)}`)
              // message+= `领取成功，获得${JSON.stringify(data.lotteryResult)}\n`
              message += `领取成功，获得 ${(data.lotteryResult.jPeasList[0].quantity)}京豆`
              allMessage += `京东账号${$.index}${$.nickName || $.UserName}\n领取成功，获得 ${(data.lotteryResult.jPeasList[0].quantity)}京豆${$.index !== cookiesArr.length ? '\n\n' : ''}`;
            } else if (data.subCode === '8') {
              console.log(`今日次数已满`)
              message += `领取失败，本场已领过`;
            } else {
              console.log(`异常：${JSON.stringify(data)}`)
            }
          }
        }
      } catch (e) {
        $.logErr(e, resp)
      } finally {
        resolve();
      }
    })
  })
}

function taskGetUrl(url, body) {
  return {
    url: url,
    body: body,
    headers: {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "zh-cn",
      "Connection": "keep-alive",
      "Content-Type": "application/x-www-form-urlencoded",
      "Host": "api.m.jd.com",
      "Referer": `https://h5.m.jd.com/active/redrain/index.html?id=${$.activityId}&lng=0.000000&lat=0.000000&sid=&un_area=`,
      "Cookie": cookie,
      "User-Agent": "JD4iPhone/9.3.5 CFNetwork/1209 Darwin/20.2.0"
    }
  }
}

function taskPostUrl(function_id, body = body) {
  return {
    url: `https://api.m.jd.com/client.action?functionId=${function_id}`,
    body: body,
    headers: {
      'Host': 'api.m.jd.com',
      'content-type': 'application/x-www-form-urlencoded',
      'accept': '*/*',
      'user-agent': 'JD4iPhone/167408 (iPhone; iOS 14.2; Scale/3.00)',
      'accept-language': 'zh-Hans-JP;q=1, en-JP;q=0.9, zh-Hant-TW;q=0.8, ja-JP;q=0.7, en-US;q=0.6',
      //"Cookie": cookie,
    }
  }
}

function taskUrl(function_id, body = {}) {
  return {
    url: `${JD_API_HOST}?functionId=${function_id}&body=${escape(JSON.stringify(body))}&client=wh5&clientVersion=1.0.0&_=${new Date().getTime() + new Date().getTimezoneOffset() * 60 * 1000 + 8 * 60 * 60 * 1000}`,
    headers: {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "zh-cn",
      "Connection": "keep-alive",
      "Content-Type": "application/x-www-form-urlencoded",
      "Host": "api.m.jd.com",
      "Referer": `https://h5.m.jd.com/active/redrain/index.html?id=${$.activityId}&lng=0.000000&lat=0.000000&sid=&un_area=`,
      "Cookie": cookie,
      "User-Agent": "JD4iPhone/9.3.5 CFNetwork/1209 Darwin/20.2.0"
    }
  }
}

function TotalBean() {
  return new Promise(async resolve => {
    const options = {
      "url": `https://wq.jd.com/user/info/QueryJDUserInfo?sceneval=2`,
      "headers": {
        "Accept": "application/json,text/plain, */*",
        "Content-Type": "application/x-www-form-urlencoded",
        "Accept-Encoding": "gzip, deflate, br",
        "Accept-Language": "zh-cn",
        "Connection": "keep-alive",
        "Cookie": cookie,
        "Referer": "https://wqs.jd.com/my/jingdou/my.shtml?sceneval=2",
        "User-Agent": $.isNode() ? (process.env.JD_USER_AGENT ? process.env.JD_USER_AGENT : "JD4iPhone/9.3.5 CFNetwork/1209 Darwin/20.2.0") : ($.getdata('JDUA') ? $.getdata('JDUA') : "JD4iPhone/9.3.5 CFNetwork/1209 Darwin/20.2.0")
      }
    }
    $.post(options, (err, resp, data) => {
      try {
        if (err) {
          console.log(`${JSON.stringify(err)}`)
          console.log(`${$.name} API请求失败，请检查网路重试`)
        } else {
          if (data) {
            data = JSON.parse(data);
            if (data['retcode'] === 13) {
              $.isLogin = false; //cookie过期
              return
            }
            if (data['retcode'] === 0) {
              $.nickName = (data['base'] && data['base'].nickname) || $.UserName;
            } else {
              $.nickName = $.UserName
            }
          } else {
            console.log(`京东服务器返回空数据`)
          }
        }
      } catch (e) {
        $.logErr(e, resp)
      } finally {
        resolve();
      }
    })
  })
}

function safeGet(data) {
  try {
    if (typeof JSON.parse(data) == "object") {
      return true;
    }
  } catch (e) {
    console.log(e);
    console.log(`京东服务器访问数据为空，请检查自身设备网络情况`);
    return false;
  }
}

function jsonParse(str) {
  if (typeof str == "string") {
    try {
      return JSON.parse(str);
    } catch (e) {
      console.log(e);
      $.msg($.name, '', '不要在BoxJS手动复制粘贴修改cookie')
      return [];
    }
  }
}
// prettier-ignore
function Env(t,e){"undefined"!=typeof process&&JSON.stringify(process.env).indexOf("GITHUB")>-1&&process.exit(0);class s{constructor(t){this.env=t}send(t,e="GET"){t="string"==typeof t?{url:t}:t;let s=this.get;return"POST"===e&&(s=this.post),new Promise((e,i)=>{s.call(this,t,(t,s,r)=>{t?i(t):e(s)})})}get(t){return this.send.call(this.env,t)}post(t){return this.send.call(this.env,t,"POST")}}return new class{constructor(t,e){this.name=t,this.http=new s(this),this.data=null,this.dataFile="box.dat",this.logs=[],this.isMute=!1,this.isNeedRewrite=!1,this.logSeparator="\n",this.startTime=(new Date).getTime(),Object.assign(this,e),this.log("",`🔔${this.name}, 开始!`)}isNode(){return"undefined"!=typeof module&&!!module.exports}isQuanX(){return"undefined"!=typeof $task}isSurge(){return"undefined"!=typeof $httpClient&&"undefined"==typeof $loon}isLoon(){return"undefined"!=typeof $loon}toObj(t,e=null){try{return JSON.parse(t)}catch{return e}}toStr(t,e=null){try{return JSON.stringify(t)}catch{return e}}getjson(t,e){let s=e;const i=this.getdata(t);if(i)try{s=JSON.parse(this.getdata(t))}catch{}return s}setjson(t,e){try{return this.setdata(JSON.stringify(t),e)}catch{return!1}}getScript(t){return new Promise(e=>{this.get({url:t},(t,s,i)=>e(i))})}runScript(t,e){return new Promise(s=>{let i=this.getdata("@chavy_boxjs_userCfgs.httpapi");i=i?i.replace(/\n/g,"").trim():i;let r=this.getdata("@chavy_boxjs_userCfgs.httpapi_timeout");r=r?1*r:20,r=e&&e.timeout?e.timeout:r;const[o,h]=i.split("@"),n={url:`http://${h}/v1/scripting/evaluate`,body:{script_text:t,mock_type:"cron",timeout:r},headers:{"X-Key":o,Accept:"*/*"}};this.post(n,(t,e,i)=>s(i))}).catch(t=>this.logErr(t))}loaddata(){if(!this.isNode())return{};{this.fs=this.fs?this.fs:require("fs"),this.path=this.path?this.path:require("path");const t=this.path.resolve(this.dataFile),e=this.path.resolve(process.cwd(),this.dataFile),s=this.fs.existsSync(t),i=!s&&this.fs.existsSync(e);if(!s&&!i)return{};{const i=s?t:e;try{return JSON.parse(this.fs.readFileSync(i))}catch(t){return{}}}}}writedata(){if(this.isNode()){this.fs=this.fs?this.fs:require("fs"),this.path=this.path?this.path:require("path");const t=this.path.resolve(this.dataFile),e=this.path.resolve(process.cwd(),this.dataFile),s=this.fs.existsSync(t),i=!s&&this.fs.existsSync(e),r=JSON.stringify(this.data);s?this.fs.writeFileSync(t,r):i?this.fs.writeFileSync(e,r):this.fs.writeFileSync(t,r)}}lodash_get(t,e,s){const i=e.replace(/\[(\d+)\]/g,".$1").split(".");let r=t;for(const t of i)if(r=Object(r)[t],void 0===r)return s;return r}lodash_set(t,e,s){return Object(t)!==t?t:(Array.isArray(e)||(e=e.toString().match(/[^.[\]]+/g)||[]),e.slice(0,-1).reduce((t,s,i)=>Object(t[s])===t[s]?t[s]:t[s]=Math.abs(e[i+1])>>0==+e[i+1]?[]:{},t)[e[e.length-1]]=s,t)}getdata(t){let e=this.getval(t);if(/^@/.test(t)){const[,s,i]=/^@(.*?)\.(.*?)$/.exec(t),r=s?this.getval(s):"";if(r)try{const t=JSON.parse(r);e=t?this.lodash_get(t,i,""):e}catch(t){e=""}}return e}setdata(t,e){let s=!1;if(/^@/.test(e)){const[,i,r]=/^@(.*?)\.(.*?)$/.exec(e),o=this.getval(i),h=i?"null"===o?null:o||"{}":"{}";try{const e=JSON.parse(h);this.lodash_set(e,r,t),s=this.setval(JSON.stringify(e),i)}catch(e){const o={};this.lodash_set(o,r,t),s=this.setval(JSON.stringify(o),i)}}else s=this.setval(t,e);return s}getval(t){return this.isSurge()||this.isLoon()?$persistentStore.read(t):this.isQuanX()?$prefs.valueForKey(t):this.isNode()?(this.data=this.loaddata(),this.data[t]):this.data&&this.data[t]||null}setval(t,e){return this.isSurge()||this.isLoon()?$persistentStore.write(t,e):this.isQuanX()?$prefs.setValueForKey(t,e):this.isNode()?(this.data=this.loaddata(),this.data[e]=t,this.writedata(),!0):this.data&&this.data[e]||null}initGotEnv(t){this.got=this.got?this.got:require("got"),this.cktough=this.cktough?this.cktough:require("tough-cookie"),this.ckjar=this.ckjar?this.ckjar:new this.cktough.CookieJar,t&&(t.headers=t.headers?t.headers:{},void 0===t.headers.Cookie&&void 0===t.cookieJar&&(t.cookieJar=this.ckjar))}get(t,e=(()=>{})){t.headers&&(delete t.headers["Content-Type"],delete t.headers["Content-Length"]),this.isSurge()||this.isLoon()?(this.isSurge()&&this.isNeedRewrite&&(t.headers=t.headers||{},Object.assign(t.headers,{"X-Surge-Skip-Scripting":!1})),$httpClient.get(t,(t,s,i)=>{!t&&s&&(s.body=i,s.statusCode=s.status),e(t,s,i)})):this.isQuanX()?(this.isNeedRewrite&&(t.opts=t.opts||{},Object.assign(t.opts,{hints:!1})),$task.fetch(t).then(t=>{const{statusCode:s,statusCode:i,headers:r,body:o}=t;e(null,{status:s,statusCode:i,headers:r,body:o},o)},t=>e(t))):this.isNode()&&(this.initGotEnv(t),this.got(t).on("redirect",(t,e)=>{try{if(t.headers["set-cookie"]){const s=t.headers["set-cookie"].map(this.cktough.Cookie.parse).toString();s&&this.ckjar.setCookieSync(s,null),e.cookieJar=this.ckjar}}catch(t){this.logErr(t)}}).then(t=>{const{statusCode:s,statusCode:i,headers:r,body:o}=t;e(null,{status:s,statusCode:i,headers:r,body:o},o)},t=>{const{message:s,response:i}=t;e(s,i,i&&i.body)}))}post(t,e=(()=>{})){if(t.body&&t.headers&&!t.headers["Content-Type"]&&(t.headers["Content-Type"]="application/x-www-form-urlencoded"),t.headers&&delete t.headers["Content-Length"],this.isSurge()||this.isLoon())this.isSurge()&&this.isNeedRewrite&&(t.headers=t.headers||{},Object.assign(t.headers,{"X-Surge-Skip-Scripting":!1})),$httpClient.post(t,(t,s,i)=>{!t&&s&&(s.body=i,s.statusCode=s.status),e(t,s,i)});else if(this.isQuanX())t.method="POST",this.isNeedRewrite&&(t.opts=t.opts||{},Object.assign(t.opts,{hints:!1})),$task.fetch(t).then(t=>{const{statusCode:s,statusCode:i,headers:r,body:o}=t;e(null,{status:s,statusCode:i,headers:r,body:o},o)},t=>e(t));else if(this.isNode()){this.initGotEnv(t);const{url:s,...i}=t;this.got.post(s,i).then(t=>{const{statusCode:s,statusCode:i,headers:r,body:o}=t;e(null,{status:s,statusCode:i,headers:r,body:o},o)},t=>{const{message:s,response:i}=t;e(s,i,i&&i.body)})}}time(t,e=null){const s=e?new Date(e):new Date;let i={"M+":s.getMonth()+1,"d+":s.getDate(),"H+":s.getHours(),"m+":s.getMinutes(),"s+":s.getSeconds(),"q+":Math.floor((s.getMonth()+3)/3),S:s.getMilliseconds()};/(y+)/.test(t)&&(t=t.replace(RegExp.$1,(s.getFullYear()+"").substr(4-RegExp.$1.length)));for(let e in i)new RegExp("("+e+")").test(t)&&(t=t.replace(RegExp.$1,1==RegExp.$1.length?i[e]:("00"+i[e]).substr((""+i[e]).length)));return t}msg(e=t,s="",i="",r){const o=t=>{if(!t)return t;if("string"==typeof t)return this.isLoon()?t:this.isQuanX()?{"open-url":t}:this.isSurge()?{url:t}:void 0;if("object"==typeof t){if(this.isLoon()){let e=t.openUrl||t.url||t["open-url"],s=t.mediaUrl||t["media-url"];return{openUrl:e,mediaUrl:s}}if(this.isQuanX()){let e=t["open-url"]||t.url||t.openUrl,s=t["media-url"]||t.mediaUrl;return{"open-url":e,"media-url":s}}if(this.isSurge()){let e=t.url||t.openUrl||t["open-url"];return{url:e}}}};if(this.isMute||(this.isSurge()||this.isLoon()?$notification.post(e,s,i,o(r)):this.isQuanX()&&$notify(e,s,i,o(r))),!this.isMuteLog){let t=["","==============📣系统通知📣=============="];t.push(e),s&&t.push(s),i&&t.push(i),console.log(t.join("\n")),this.logs=this.logs.concat(t)}}log(...t){t.length>0&&(this.logs=[...this.logs,...t]),console.log(t.join(this.logSeparator))}logErr(t,e){const s=!this.isSurge()&&!this.isQuanX()&&!this.isLoon();s?this.log("",`❗️${this.name}, 错误!`,t.stack):this.log("",`❗️${this.name}, 错误!`,t)}wait(t){return new Promise(e=>setTimeout(e,t))}done(t={}){const e=(new Date).getTime(),s=(e-this.startTime)/1e3;this.log("",`🔔${this.name}, 结束! 🕛 ${s} 秒`),this.log(),(this.isSurge()||this.isQuanX()||this.isLoon())&&$done(t)}}(t,e)}
EOF

echo "${lrr//RRAID//$1}" >/ql/scripts/jrd1.js

echo "红包雨RRA替换成$1完成"

# 删除旧的定时任务
#sed -i '/jrd1.js/d' /jd/config/crontab.list

#echo "删除旧的jrd1定时任务完成"

# 添加今天的定时任务
# echo "1 0-23$(date "+ %-d %-m") * jtask jrd1.js now"  >>  /jd/config/crontab.list

# echo "添加当天的jrd1红包雨定时任务完成"

#echo "顺便白嫖一下红包雨"

#task jrd1.js now
